CoThings/ViewController.swift

```diff
@@ -332,3 +325,46 @@ class ViewController: UIViewController, CLLocationManagerDelegate, UITableViewDe
         print("You tapped cell number \(indexPath.row).")
     }
 }
+
+
+
+class MessageAdapter {
+    
+    struct RoomsResponse: Decodable {
+        enum CodingKeys: String, CodingKey {
+            case status
+            case response
+        }
+        
+        enum ResponseKeys: String, CodingKey {
+            case rooms
+        }
+        
+        var status: String
+        var rooms: [Room]
+        
+        init(from decoder: Decoder) throws {
+            let container = try decoder.container(keyedBy: CodingKeys.self)
+            status = try container.decode(String.self, forKey: .status)
+            let nestedContainer = try container.nestedContainer(keyedBy: ResponseKeys.self, forKey: .response)
+            rooms = try nestedContainer.decode([Room].self, forKey: .rooms)
+        }
+    }
+    
+    private var message: Message
+    
+    init(message: Message) {
+        self.message = message
+    }
+    
+    func decodeRoomResponse() -> RoomsResponse? {
+        do {
+            let jsonData = try JSONSerialization.data(withJSONObject: message.payload, options: .fragmentsAllowed)
+            let decoder = JSONDecoder()
+            let roomResponse = try decoder.decode(RoomsResponse.self, from: jsonData)
+            return roomResponse
+        } catch {
+            return nil
```

Comment:
Feels like it's better to let it throw, and handle the error on the caller side.

Reason: the input provided here is obviously wrong/faulty.

Reply:
if we are throwing an error I think we probably be throwing a different error instead of the one thrown by decoding error.  like conversion error of some sort

Reply:
I agree, that makes more sense.

Reply:
I think last update also covers this, resolving!